### PR TITLE
CUDA CSR diagonal kernels

### PR DESCRIPTION
**Pull request by cnpetra**
Provides a couple of kernels for updating diagonal of a CSR matrix

### CODE REVIEW

File: src/LinAlg/MatrixSparseCsrCudaKernels.cu [hunk 1]
```diff
@@ -63,7 +63,7 @@
 #include <device_launch_parameters.h>
 
 __global__
-void csr_set_diag_to_val(int n, int nnz, int* irowptr, int* jcolind, double* values, double val)
+void csr_add_vec_to_diag(int n, int nnz, int* irowptr, int* jcolind, double* values, double alpha, const double* diag_values)
```
Comment (nychiang) [inline src/LinAlg/MatrixSparseCsrCudaKernels.cu:66]: I think it's better to add some descriptions to each function in this file.

File: src/LinAlg/MatrixSparseCsrCudaKernels.hpp [hunk 2]
```diff
@@ -62,9 +62,47 @@ namespace cuda
...
+ * for each row. Assumes pointers are on the device and parallelizes over rows.
+ * 
+ * @pre CSR matrix must be square.
+ * @pre Diagonal entries must appear explicitly among the nonzeros.
+ * @pre Column indexes must be sorted for any given row.
+ */   
+void csr_add_diag_kernel(int n, int nnz, int* irowptr, int* jcoldind, double* values, double Dval);
+
+/**
+ * Add entries of the array `values`  the diagonal of the CSR matrix. Performs a binary search on the column indexes
```
Comment (nychiang) [inline src/LinAlg/MatrixSparseCsrCudaKernels.hpp:83]: Add entries of the array `values` "to" the diagonal of the CSR matrix.

File: src/LinAlg/hiopMatrixSparseCsrCuda.cpp [hunk 3]
```diff
@@ -773,7 +779,7 @@ void hiopMatrixSparseCSRCUDA::form_transpose_from_numeric(const hiopMatrixSparse
 
 void hiopMatrixSparseCSRCUDA::form_diag_from_symbolic(const hiopVector& D)
 {
-  int m = D.get_size();
+  const int m = D.get_size();
```
Comment (nychiang) [inline src/LinAlg/hiopMatrixSparseCsrCuda.cpp:782]: Is this function finished? seems that the sparsity patten hasn't been assigned yet.

**Review by nychiang — APPROVED** [review]
@cnpetra  by using the new functions "csr_add_diag_kernel" and "csr_set_diag_kernel", can we delete the diagonal matrix i[n this section?](https://github.com/LLNL/hiop/blob/54fba5c1d11fd2f5c56d46bdb039d2583fa6bca3/src/Optimization/hiopKKTLinSysSparseNormalEqn.cpp#L241-L257)

**Comment by cnpetra** [comment]
> @cnpetra by using the new functions "csr_add_diag_kernel" and "csr_set_diag_kernel", can we delete the diagonal matrix i[n this section?](https://github.com/LLNL/hiop/blob/54fba5c1d11fd2f5c56d46bdb039d2583fa6bca3/src/Optimization/hiopKKTLinSysSparseNormalEqn.cpp#L241-L257)

Probably yes, but not sure which one is faster. I will follow up later on once I can benchmark the code a bit more.

File: src/LinAlg/hiopMatrixSparseCsrCuda.cpp (hunk 3 above)
Comment (cnpetra) [reply src/LinAlg/hiopMatrixSparseCsrCuda.cpp:782]: it is not, that code got lost somewhere somehow ...

**Review by nychiang — APPROVED** [review]

**Pull request merged**